pkg/operator/certrotation/signer_test.go

```diff
@@ -1,24 +1,219 @@
+		// which could result in a 404 thus
+		// to simulate this scenario
+		// the secret si removed from the cache
+		if err = indexer.Delete(existingSecret); err != nil {
+			return true, nil, err
+		}
+		close(secondControllerStartChan) // this triggers the second controller
+		return false /*means that the next reactor in the chain will be invoked (secret will be deleted)*/, nil, nil
+	})
+	fakeKubeClient.PrependReactor("get", "secrets", func() clienttesting.ReactionFunc {
+		counter := 0
+		counterLock := sync.Mutex{}
+		return func(action clienttesting.Action) (bool, runtime.Object, error) {
+			counterLock.Lock()
+			defer counterLock.Unlock()
+			defer func() {
+				counter++
+			}()
+			if counter == 0 {
+				// this is the call from the first controller
+				return true, existingSecret, nil
+			}
+			// this is a call from the second controller
+			// wait until the first controller finishes
+			// so that we can update the secret with the new keys
+			<-firstControllerEndChan
+
+			// cannot use fakeKubeClient.ActionsA()
+			// because it deadlocks, fakeKubeClient.RLock already acquired
+			//
+			// the created secret from the actions
+			// is validated at the end of this test
+			// and must match this one
+			createdSecret := existingSecret.DeepCopy()
+			createdSecret.Type = corev1.SecretTypeTLS
+			createdSecret.Annotations["openshift.io/owning-component"] = "test"
+			return true, createdSecret, nil
+		}
+	}())
+
+	c := &RotatedSigningCASecret{
+		Namespace:     "ns",
+		Name:          "signer",
+		Validity:      24 * time.Hour,
+		Refresh:       12 * time.Hour,
+		Client:        fakeKubeClient.CoreV1(),
+		Lister:        corev1listers.NewSecretLister(indexer),
+		EventRecorder: events.NewInMemoryRecorder("test"),
+		AdditionalAnnotations: AdditionalAnnotations{
+			JiraComponent: "test",
+		},
+		Owner: &metav1.OwnerReference{
+			Name: "operator",
+		},
+	}
+
+	var signingKeyPairFromControllerTwo *crypto.CA
+	go func() {
+		// start the second controller
+		var err error
+		<-secondControllerStartChan
+		signingKeyPairFromControllerTwo, err = c.ensureSigningCertKeyPair(context.TODO())
+		if err != nil {
+			t.Error(err)
+		}
+		close(secondControllerEndChan)
+	}()
+
+	// start the first controller
+	signingKeyPairFromControllerOne, err := c.ensureSigningCertKeyPair(context.TODO())
+	if err != nil {
+		t.Fatal(err)
+	}
+	certBytesFromControllerOne, keyBytesFromControllerOne, err := signingKeyPairFromControllerOne.Config.GetPEMBytes()
+	if err != nil {
+		t.Fatal(err)
+	}
+	if !bytes.Equal(existingSecret.Data["tls.crt"], certBytesFromControllerOne) {
+		t.Fatal("returned certificate data from controller one doesn't match existing secreted data")
+	}
+	if !bytes.Equal(existingSecret.Data["tls.key"], keyBytesFromControllerOne) {
+		t.Fatal("returned certificate key from controller one doesn't match existing secreted data")
+	}
+	close(firstControllerEndChan)
+	<-secondControllerEndChan
```

Comment:
In general, it might be a good idea to have a timeout on all of these channel receive statements. A change to the one of the functions covered by this test could make the test itself timeout or deadlock instead of failing in an obvious way.